primitives/forest-verifier/src/tests.rs

```diff
@@ -223,1153 +226,1524 @@ pub fn merklise_file<T: TrieLayout>(file_path: &str) -> (MemoryDB<T::Hash>, Hash
 
-    {
-        // Creating trie inside of closure to drop it before generating proof.
-        let mut trie_recorder = recorder.as_trie_recorder(root);
-        let trie = TrieDBBuilder::<LayoutV1<BlakeTwo256>>::new(&memdb, &root)
-            .with_recorder(&mut trie_recorder)
-            .build();
-
-        // Create an iterator over the leaf nodes.
-        let mut iter = trie.into_double_ended_iter().unwrap();
-
-        // Seek to the challenge key, this will only look up the next leaf node.
-        iter.seek(&challenge_key.0).unwrap();
-        iter.next_back().unwrap().unwrap();
-
-        // Seek to two keys ahead of the challenge key.
-        iter.seek(&next_leaf_key.0).unwrap();
-        iter.next().unwrap().unwrap();
+        let err =
+            match ForestVerifier::<LayoutV1<BlakeTwo256>, { BlakeTwo256::LENGTH }>::apply_delta(
+                &root, &mutations, &proof,
+            ) {
+                Err(err) => err,
+                Ok(_) => panic!("Expected error."),
+            };
+
+        assert_eq!(err, DispatchError::Other("No mutations provided."));
     }
 
-    // Generate proof
-    let proof = recorder
-        .drain_storage_proof()
-        .to_compact_proof::<BlakeTwo256>(root)
-        .expect("Failed to create compact proof from recorder");
+    #[test]
+    fn mutate_root_does_not_match_expected_fail() {
+        let (memdb, root, leaf_keys, mut recorder) = setup_trie_and_recorder();
 
-    assert_eq!(
-        ForestVerifier::<LayoutV1<BlakeTwo256>, { BlakeTwo256::LENGTH }>::verify_proof(
-            &root,
-            &[challenge_key],
-            &proof
-        ),
-        Err("Failed to get next leaf.".into())
-    );
+        let challenge_key = *leaf_keys.first().unwrap();
+        let mutations: Vec<(H256, TrieMutation)> =
+            vec![(challenge_key, TrieRemoveMutation::default().into())];
+
+        {
+            let mut trie_recorder = recorder.as_trie_recorder(root);
+            let trie = TrieDBBuilder::<LayoutV1<BlakeTwo256>>::new(&memdb, &root)
+                .with_recorder(&mut trie_recorder)
+                .build();
+
+            let mut iter = trie.into_double_ended_iter().unwrap();
+            iter.seek(&challenge_key.0).unwrap();
+
+            // Access the next leaf node.
+            iter.next();
+        }
+
+        let proof = generate_proof_and_verify(&mut recorder, &root, &[challenge_key]);
+
+        let err =
+            match ForestVerifier::<LayoutV1<BlakeTwo256>, { BlakeTwo256::LENGTH }>::apply_delta(
+                &H256([0; 32]),
+                &mutations,
+                &proof,
+            ) {
+                Err(err) => err,
+                Ok(_) => panic!("Expected error."),
+            };
+
+        assert_eq!(
+            err,
+            DispatchError::Other(
+                "Failed to convert proof to memory DB, root doesn't match with expected."
+            )
+        );
```

Comment:
I think you could have used `assert_err!` macro for this.

Reply:
Ah thank you!